File: tools/lycon-browser/windows/LyconWindows/LyconBridge.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Text.Json;
using System.Threading.Tasks;
using System.Windows;
using Microsoft.Web.WebView2.Core;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;

namespace LyconWindows;

/// <summary>
/// Implements the Lycon bridge contract on WinUI 3 + WebView2.
/// Receives invoke() calls from JS via postMessage and routes them to handlers.
/// </summary>
public class LyconBridge
{
    private readonly LyconDataService _data;
    private readonly LyconShieldsService _shields;
    private readonly MainWindow _window;
    private readonly Dictionary<string, Func<JToken?, Task<object>>> _handlers;
    private int _nextCallId = 1;
    private readonly Dictionary<int, TaskCompletionSource<object>> _pendingJsCalls = new();

    public LyconBridge(LyconDataService data, LyconShieldsService shields, MainWindow window)
    {
        _data = data;
        _shields = shields;
        _window = window;

        // ----- Search engines (mirror Electron main.js) -----
        var searchEngines = new Dictionary<string, object>
        {
            ["brave"] = new { name = "Brave Search", url = "https://search.brave.com/search?q=%s", suggest = "https://search.brave.com/api/suggest?q=%s" },
            ["duckduckgo"] = new { name = "DuckDuckGo", url = "https://duckduckgo.com/?q=%s", suggest = "https://duckduckgo.com/ac/?q=%s" },
            ["google"] = new { name = "Google", url = "https://www.google.com/search?q=%s", suggest = "https://suggestqueries.google.com/complete/search?client=firefox&q=%s" },
            ["bing"] = new { name = "Bing", url = "https://www.bing.com/search?q=%s", suggest = "https://www.bing.com/osjson.aspx?query=%s" },
            ["startpage"] = new { name = "Startpage", url = "https://www.startpage.com/sp/search?query=%s", suggest = "https://www.startpage.com/sp/suggest?q=%s" },
        };

        _handlers = new Dictionary<string, Func<JToken?, Task<object>>>
        {
            ["settings:get"] = _ => Task.FromResult<object>(_data.LoadSettings()),
            ["settings:set"] = payload =>
            {
                var patch = payload?.ToObject<LyconSettings>() ?? new LyconSettings();
                return Task.FromResult<object>(_data.UpdateSettings(patch));
            },
            ["search:list"] = _ => Task.FromResult<object>(searchEngines),
            ["search:build"] = payload =>
            {
                var args = payload?.ToObject<SearchBuildArgs>() ?? new SearchBuildArgs();
                var eng = searchEngines.TryGetValue(args.Engine ?? "brave", out var e) ? e : searchEngines["brave"];
                var urlProp = ((dynamic)e).url as string;
                return Task.FromResult<object>(urlProp.Replace("%s", Uri.EscapeDataString(args.Query ?? "")));
            },
            ["bookmarks:list"] = _ => Task.FromResult<object>(_data.LoadBookmarks()),
            ["bookmarks:add"] = payload =>
            {
                var bm = payload?.ToObject<LyconBookmark>() ?? new LyconBookmark();
                return Task.FromResult<object>(_data.AddBookmark(bm));
            },
            ["bookmarks:remove"] = payload => Task.FromResult<object>(_data.RemoveBookmark(payload?.ToObject<string>() ?? "")),
            ["history:list"] = _ => Task.FromResult<object>(_data.LoadHistory()),
            ["history:add"] = payload =>
            {
                var entry = payload?.ToObject<LyconHistoryEntry>() ?? new LyconHistoryEntry();
                return Task.FromResult<object>(_data.AddHistory(entry));
            },
            ["history:remove"] = payload => Task.FromResult<object>(_data.RemoveHistory(payload?.ToObject<string>() ?? "")),
            ["history:clear"] = _ => { _data.ClearHistory(); return Task.FromResult<object>(new object[0]); },
            ["downloads:list"] = _ => Task.FromResult<object>(_data.LoadDownloads()),
            ["downloads:open"] = payload => { Process.Start("explorer.exe", payload?.ToObject<string>() ?? ""); return Task.FromResult<object>(true); },
            ["downloads:show"] = payload => { Process.Start("explorer.exe", "/select," + (payload?.ToObject<string>() ?? "")); return Task.FromResult<object>(true); },
            ["downloads:clear"] = _ => { _data.ClearDownloads(); return Task.FromResult<object>(new object[0]); },
            ["shields:toggle"] = payload =>
            {
                var enabled = payload?.ToObject<bool>() ?? true;
                _shields.IsEnabled = enabled;
                var s = _data.LoadSettings(); s.ShieldsEnabled = enabled; _data.SaveSettings(s);
                return Task.FromResult<object>(s);
            },
            ["shields:status"] = _ => Task.FromResult<object>(new { enabled = _shields.IsEnabled, totalBlocked = _shields.TotalBlocked, blockerLoaded = true }),
            ["window:minimize"] = _ => { _window.Minimize(); return Task.FromResult<object>(true); },
            ["window:maximize"] = _ => { _window.Maximize(); return Task.FromResult<object>(true); },
            ["window:close"] = _ => { _window.Close(); return Task.FromResult<object>(true); },
            ["shell:openExternal"] = payload =>
            {
                var url = payload?.ToObject<string>() ?? "";
                if (!string.IsNullOrEmpty(url)) Process.Start(new ProcessStartInfo(url) { UseShellExecute = true });
                return Task.FromResult<object>(true);
            },
        };
    }

    /// <summary>
    /// Generates the JS source that creates window.__lyconNative on the page.
    /// The script uses postMessage for invoke() and listens for messages
    /// from native for both invoke responses and events.
    /// </summary>
    public string GetBridgeInitScript()
    {
        return @"
(function() {
  if (window.__lyconNative) return; // already initialized
  const pending = new Map(); // callId -> { resolve, reject }
  let nextCallId = 1;
  const eventListeners = new Map(); // event -> Set<cb>

  function sendToNative(msg) {
    window.chrome.webview.postMessage(JSON.stringify(msg));
  }

  window.chrome.webview.addEventListener('message', function(e) {
    let data;
    try { data = JSON.parse(e.data); } catch(_) { return; }
    if (data.type === 'lycon:response') {
      const p = pending.get(data.callId);
      if (p) {
        pending.delete(data.callId);
        if (data.error) p.reject(new Error(data.error));
        else p.resolve(data.result);
      }
    } else if (data.type === 'lycon:event') {
      const set = eventListeners.get(data.event);
      if (set) for (const cb of set) {
        try { cb(data.payload); } catch(err) { console.error('[Lycon bridge] listener error', err); }
      }
    }
  });

  window.__lyconNative = {
    invoke: function(action, payload) {
      return new Promise((resolve, reject) => {
        const callId = nextCallId++;
        pending.set(callId, { resolve, reject });
        sendToNative({ type: 'lycon:invoke', callId, action, payload: payload ?? null });
      });
    },
    on: function(event, cb) {
      if (!eventListeners.has(event)) eventListeners.set(event, new Set());
      eventListeners.get(event).add(cb);
      return function() {
        const set = eventListeners.get(event);
        if (set) set.delete(cb);
      };
    },
    platform: 'winui',
    versions: { webview2: '1.0.2739.15', os: 'Windows' },
    initialUrl: null,
  };
  console.log('[Lycon] WinUI bridge initialized');
})();
";
    }

    /// <summary>
    /// Handles a JSON message received from JS via WebMessageReceived.
    /// </summary>
    public async void HandleMessageFromJs(string json)
    {
        try
        {
            var msg = JObject.Parse(json);
            var type = msg["type"]?.ToString();
            if (type != "lycon:invoke")
            {
                Debug.WriteLine($"[Lycon] Unknown message type: {type}");
                return;
            }

            var callId = msg["callId"]?.Value<int>() ?? 0;
            var action = msg["action"]?.ToString() ?? "";
            var payload = msg["payload"];

            try
            {
                if (!_handlers.TryGetValue(action, out var handler))
                    throw new Exception($"No handler registered for action '{action}'");

                var result = await handler(payload);
                SendResponse(callId, result, error: null);
            }
            catch (Exception ex)
            {
                SendResponse(callId, null, error: ex.Message);
            }
        }
        catch (Exception ex)
        {
            Debug.WriteLine($"[Lycon] HandleMessageFromJs error: {ex.Message}");
        }
    }

    private void SendResponse(int callId, object? result, string? error)
    {
        var json = System.Text.Json.JsonSerializer.Serialize(new
        {
            type = "lycon:response",
            callId,
            result,
            error,
        });
        _window.SendEventToJsRaw(json);
    }

    /// <summary>
    /// Pushes an event to the JS side.
    /// </summary>
    public void SendEvent(string eventType, object payload)
    {
        _window.SendEventToJs(eventType, payload);
    }
}

internal class SearchBuildArgs
{
    [JsonProperty("engine")] public string? Engine { get; set; }
    [JsonProperty("query")] public string? Query { get; set; }
}

File: tools/lycon-browser/android/app/src/main/java/com/lycon/browser/LyconBridge.kt
package com.lycon.browser

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.util.Log
import org.json.JSONArray
import org.json.JSONObject
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicInteger

/**
 * Bridge between GeckoView's JS and native Kotlin code.
 *
 * Protocol:
 *   JS -> Native:  GeckoSession prompt delegate intercepts window.prompt() calls
 *                  of the form: lycon:invoke:<callId>:<action>:<payloadJson>
 *                  (Using prompt() is the standard GeckoView bridge pattern.)
 *
 *   Native -> JS:  GeckoSession.evaluateJavaScript("window.__lyconBridge.onResponse(...)")
 *                  and "window.__lyconBridge.onEvent(...)"
 *
 * The bridge script (injected by MainActivity) creates window.__lyconNative
 * which calls window.prompt() with the encoded message.
 */
class LyconBridge(
    private val context: Context,
    private val dataService: LyconDataService,
    private val shieldsService: LyconShieldsService,
    private val onEventSender: (String, JSONObject) -> Unit,
    private val onWindowClose: () -> Unit,
    private val onWindowMinimize: () -> Unit,
    private val onWindowMaximize: () -> Unit,
) {

    private val handlers: Map<String, (JSONObject?) -> Any> = mapOf(
        "settings:get" to { _ -> dataService.loadSettings() },
        "settings:set" to { payload ->
            val patch = payload ?: JSONObject()
            dataService.updateSettings(patch)
        },
        "search:list" to { _ -> SEARCH_ENGINES },
        "search:build" to { payload ->
            val engine = payload?.optString("engine") ?: "brave"
            val query = payload?.optString("query") ?: ""
            val eng = SEARCH_ENGINES.optJSONObject(engine) ?: SEARCH_ENGINES.optJSONObject("brave")!!
            eng.optString("url").replace("%s", Uri.encode(query))
        },
        "bookmarks:list" to { _ -> dataService.listBookmarks() },
        "bookmarks:add" to { payload -> dataService.addBookmark(payload ?: JSONObject()) },
        "bookmarks:remove" to { payload ->
            dataService.removeBookmark(payload?.toString() ?: "")
        },
        "history:list" to { _ -> dataService.listHistory() },
        "history:add" to { payload -> dataService.addHistory(payload ?: JSONObject()) },
        "history:remove" to { payload -> dataService.removeHistory(payload?.toString() ?: "") },
        "history:clear" to { _ -> dataService.clearHistory(); JSONArray() },
        "downloads:list" to { _ -> dataService.listDownloads() },
        "downloads:open" to { payload ->
            val path = payload?.toString() ?: ""
            val intent = Intent(Intent.ACTION_VIEW).apply {
                setDataAndType(Uri.fromFile(java.io.File(path)), "*/*")
                addFlags(Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_GRANT_READ_URI_PERMISSION)
            }
            context.startActivity(Intent.createChooser(intent, "Open").apply {
                addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            })
            true
        },
        "downloads:show" to { _ -> true /* Android doesn't have a Finder equivalent */ },
        "downloads:clear" to { _ -> dataService.clearDownloads(); JSONArray() },
        "shields:toggle" to { payload ->
            val enabled = payload?.optBoolean("enabled") ?: true
            shieldsService.isEnabled = enabled
            val s = dataService.loadSettings()
            s.put("shieldsEnabled", enabled)
            dataService.updateSettings(s)
            s
        },
        "shields:status" to { _ ->
            JSONObject().apply {
                put("enabled", shieldsService.isEnabled)
                put("totalBlocked", shieldsService.totalBlocked)
                put("blockerLoaded", shieldsService.isLoaded)
            }
        },
        "window:minimize" to { _ -> onWindowMinimize(); true },
        "window:maximize" to { _ -> onWindowMaximize(); true },
        "window:close" to { _ -> onWindowClose(); true },
        "shell:openExternal" to { payload ->
            val url = payload?.toString() ?: ""
            if (url.isNotEmpty()) {
                val intent = Intent(Intent.ACTION_VIEW, Uri.parse(url)).apply {
                    addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                }
                context.startActivity(intent)
            }
            true
        },
    )

    /**
     * Handle a JS invoke call. Returns the result synchronously.
     * Throws if the action is unknown or the handler fails.
     */
    fun handleInvoke(action: String, payload: JSONObject?): Any {
        val handler = handlers[action] ?: throw Exception("No handler for action '$action'")
        return try {
            handler(payload)
        } catch (e: Exception) {
            Log.e(TAG, "Handler error for $action", e)
            throw e
        }
    }

    /**
     * JS-side bridge source — creates window.__lyconNative via window.prompt()
     * (GeckoView's standard JS-to-native communication pattern).
     *
     * The prompt format is:  lycon:invoke:<callId>:<action>:<payloadJson>
     * Native intercepts via PromptDelegate and returns the response as the
     * prompt's return value (a JSON string).
     */
    fun getBridgeInitScript(): String = """
(function() {
  if (window.__lyconNative) return;
  const pending = new Map();
  let nextCallId = 1;

  function callNative(action, payload) {
    return new Promise((resolve, reject) => {
      const callId = nextCallId++;
      pending.set(callId, { resolve, reject });
      const msg = 'lycon:invoke:' + callId + ':' + action + ':' + (payload ? JSON.stringify(payload) : 'null');
      try {
        const response = window.prompt(msg);
        if (response === null || response === undefined) {
          pending.delete(callId);
          reject(new Error('Native returned null'));
          return;
        }
        const parsed = JSON.parse(response);
        pending.delete(callId);
        if (parsed.error) reject(new Error(parsed.error));
        else resolve(parsed.result);
      } catch (e) {
        pending.delete(callId);
        reject(e);
      }
    });
  }

  const eventListeners = new Map();

  window.__lyconNative = {
    invoke: callNative,
    on: function(event, cb) {
      if (!eventListeners.has(event)) eventListeners.set(event, new Set());
      eventListeners.get(event).add(cb);
      return function() {
        const set = eventListeners.get(event);
        if (set) set.delete(cb);
      };
    },
    platform: 'android',
    versions: { geckoview: '124.0', os: 'Android' },
    initialUrl: null,
  };

  // Native code calls this to push events
  window.__lyconBridge = {
    onEvent: function(event, payload) {
      const set = eventListeners.get(event);
      if (set) for (const cb of set) {
        try { cb(payload); } catch (e) { console.error('[Lycon bridge] listener error', e); }
      }
    }
  };

  console.log('[Lycon] Android/GeckoView bridge initialized');
})();
""".trimIndent()

    /**
     * Push an event to the JS side via evaluateJavaScript.
     */
    fun sendEvent(event: String, payload: JSONObject) {
        onEventSender(event, payload)
    }

    companion object {
        private const val TAG = "LyconBridge"

        val SEARCH_ENGINES = JSONObject().apply {
            put("brave", JSONObject().apply {
                put("name", "Brave Search")
                put("url", "https://search.brave.com/search?q=%s")
                put("suggest", "https://search.brave.com/api/suggest?q=%s")
            })
            put("duckduckgo", JSONObject().apply {
                put("name", "DuckDuckGo")
                put("url", "https://duckduckgo.com/?q=%s")
                put("suggest", "https://duckduckgo.com/ac/?q=%s")
            })
            put("google", JSONObject().apply {
                put("name", "Google")
                put("url", "https://www.google.com/search?q=%s")
                put("suggest", "https://suggestqueries.google.com/complete/search?client=firefox&q=%s")
            })
            put("bing", JSONObject().apply {
                put("name", "Bing")
                put("url", "https://www.bing.com/search?q=%s")
                put("suggest", "https://www.bing.com/osjson.aspx?query=%s")
            })
            put("startpage", JSONObject().apply {
                put("name", "Startpage")
                put("url", "https://www.startpage.com/sp/search?query=%s")
                put("suggest", "https://www.startpage.com/sp/suggest?q=%s")
            })
        }
    }
}
